modules/nlu/src/backend/pipelines/entities/pattern_extractor.ts

```diff
@@ -38,37 +40,44 @@ export default class PatternExtractor {
   ): Promise<sdk.NLU.Entity[]> {
     const values = await Promise.all(
       [occurence.name, ...occurence.synonyms].map(async x =>
-        (await this.languageProvider.tokenize(x.toLowerCase(), ds.language)).map(sanitize).filter(t => t.length)
+        (await this.languageProvider.tokenize(x.toLowerCase(), ds.language)).filter(t => t.length)
       )
     )
 
     const findings: sdk.NLU.Entity[] = []
 
-    let cur = 0
-    for (const tok of ds.tokens) {
-      cur = cur + ds.lowerText.substr(cur).indexOf(tok)
+    for (const { tok, tokenIndex } of ds.tokens.map((t, i) => ({ tok: t, tokenIndex: i }))) {
+      const rawToken = tok.value
 
       let highest = 0
       let extracted = ''
       let source = ''
+      let lastToken = tok
 
       for (const val of values) {
-        let partOfPhrase = tok
+        let partOfPhrase: string = rawToken
         const occ = val.join('+')
+        let currentLastToken = tok
 
         if (val.length > 1) {
-          const text = ds.lowerText.substr(cur + partOfPhrase.length)
-          // TODO use ds.tokens
-          const _tokens = (await this.languageProvider.tokenize(text, ds.language)).map(sanitize).filter(t => t.length)
-
-          while (_tokens && _tokens.length && partOfPhrase.length < occ.length) {
-            partOfPhrase += '+' + _tokens.shift()
+          const remainingTokens = ds.tokens.slice(tokenIndex + 1)
+
```

Comment:
Is it me or the only thing we are doing in the while is joining together the remaining tokens with a "+"?
And keeping a reference to the last?

Why not partOfPhrase = remainingTokens.map(_.get('value')).join("+")?

Reply:
Yeah, well the loop has to stop when: (partOfPhrase.length < occ.length)

Reply:
right, but we could do   const remainingTokens = ds.tokens.slice(tokenIndex + 1, occ.length).

No?

Reply:
Well no, occ is a string while tokens is an array of string so there lenght are unrelated.

There might be a way to get around the while loop, but I think it's fine as it is.

Reply:
ahhh I see 👍 right on